<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content=CCBR><link rel="shortcut icon" href=../../img/favicon.ico><title>Resources - RENEE Documentation</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Resources";
        var mkdocs_page_input_path = "RNA-seq/Resources.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> RENEE Documentation </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>About</a> </li> </ul> <ul> <li class=toctree-l1><a href=../Theory/ class="reference internal">Theory</a> </li> </ul> <ul> <li class=toctree-l1><a href=../TLDR-RNA-seq/ class="reference internal">Getting started</a> </li> </ul> <p class=caption><span class=caption-text>Command Line</span></p> <ul> <li class=toctree-l1><a href=../run/ class="reference internal">run</a> </li> <li class=toctree-l1><a href=../build/ class="reference internal">build</a> </li> <li class=toctree-l1><a href=../cache/ class="reference internal">cache</a> </li> <li class=toctree-l1><a href=../unlock/ class="reference internal">unlock</a> </li> </ul> <ul> <li class=toctree-l1><a href=../gui/ class="reference internal">Graphical Interface</a> </li> </ul> <ul> <li class=toctree-l1><a href=../output/ class="reference internal">Expected Output</a> </li> </ul> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Resources</a> <ul class=current> </ul> </li> </ul> <p class=caption><span class=caption-text>FAQ</span></p> <ul> <li class=toctree-l1><a href=../../general-questions/ class="reference internal">General Questions</a> </li> <li class=toctree-l1><a href=../../troubleshooting/ class="reference internal">Troubleshooting</a> </li> <li class=toctree-l1><a href=../../contributing/ class="reference internal">How to contribute</a> </li> </ul> <ul> <li class=toctree-l1><a href=../../license/ class="reference internal">License</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>RENEE Documentation</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Resources</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/RENEE/edit/master/docs/RNA-seq/Resources.md>Edit on CCBR/RENEE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h2 id=1-reference-genomes>1. Reference genomes<a class=headerlink href=#1-reference-genomes title="Permanent link">&para;</a></h2> <p>On <a href=https://hpc.nih.gov/ >Biowulf</a>, RENEE comes bundled with the following pre-built <a href=https://www.gencodegenes.org/ >GENCODE</a><sup>1</sup> reference genomes:</p> <p>As of RENEE v2.6.0, all hg19 and hg38 indices were built using the <a href=https://gdc.cancer.gov/about-data/gdc-data-processing/gdc-reference-files>NCI Genomic Data Commons reference fasta</a>, which contains the primary genome from Encode plus virus and decoy sequences. The hg38 fasta files were downloaded from the GDC with virus and decoy sequences already added, while these sequences were manually added to the hg19 fasta from Encode. See details here: <a href=https://github.com/CCBR/build-renee-refs>https://github.com/CCBR/build-renee-refs</a></p> <table> <thead> <tr> <th><strong>Genome</strong></th> <th><strong>Species</strong></th> <th><strong>Annotation Version</strong></th> <th><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>hg19_19</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_19.html>Gencode Release 19</a></td> <td><a href=https://www.gencodegenes.org/human/release_19.html>GRCh37</a>, Annotation Release date: 07/2013</td> </tr> <tr> <td>hg19_36</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_36lift37.html>Gencode Release 36-lift-37</a></td> <td><a href=https://www.gencodegenes.org/human/release_36lift37.html>GRCh37</a></td> </tr> <tr> <td>hg38_30</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_30.html>Gencode Release 30</a></td> <td><a href=https://www.gencodegenes.org/human/release_30.html>GRCh38</a>, Annotation Release date: 11/2018</td> </tr> <tr> <td>hg38_34</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_34.html>Gencode Release 34</a></td> <td><a href=https://www.gencodegenes.org/human/release_34.html>GRCh38</a>, Annotation Release date: 04/2020</td> </tr> <tr> <td>hg38_36</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_36.html>Gencode Release 36</a></td> <td><a href=https://www.gencodegenes.org/human/release_36.html>GRCh38</a>, Annotation Release date: 05/2020</td> </tr> <tr> <td>hg38_38</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_38.html>Gencode Release 38</a></td> <td><a href=https://www.gencodegenes.org/human/release_38.html>GRCh38</a>, Annotation Release date: 05/2021</td> </tr> <tr> <td>hg38_41</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_41.html>Gencode Release 41</a></td> <td><a href=https://www.gencodegenes.org/human/release_41.html>GRCh38</a>, Annotation Release date: 07/2022</td> </tr> <tr> <td>hg38_45</td> <td>Homo sapiens (human)</td> <td><a href=https://www.gencodegenes.org/human/release_45.html>Gencode Release 45</a></td> <td><a href=https://www.gencodegenes.org/human/release_45.html>GRCh38</a>, Annotation Release date: 03/2023</td> </tr> <tr> <td>mm10_M21</td> <td>Mus musculus (mouse)</td> <td><a href=https://www.gencodegenes.org/mouse/release_M21.html>Gencode Release M21</a></td> <td><a href=https://www.gencodegenes.org/mouse/release_M21.html>GRCm38</a>, Annotation Release date: 04/2019</td> </tr> <tr> <td>mm10_M23</td> <td>Mus musculus (mouse)</td> <td><a href=https://www.gencodegenes.org/mouse/release_M23.html>Gencode Release M23</a></td> <td><a href=https://www.gencodegenes.org/mouse/release_M23.html>GRCm38</a>, Annotation Release date: 09/2019</td> </tr> <tr> <td>mm10_M25</td> <td>Mus musculus (mouse)</td> <td><a href=https://www.gencodegenes.org/mouse/release_M25.html>Gencode Release M25</a></td> <td><a href=https://www.gencodegenes.org/mouse/release_M25.html>GRCm38</a>, Annotation Release date: 04/2020</td> </tr> <tr> <td>mCalJac1_2021</td> <td>Callithrix jacchus (white-tufted-ear marmoset)</td> <td><a href=https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_011100555.1/ >Genome assembly mCalJa1.2.pat.X</a></td> <td>Annotation release date: 04/2021</td> </tr> </tbody> </table> <p>You can run <code>renee run --help</code> to view the most up-to-date list of genome annotations available in your installation of RENEE.</p> <blockquote> <p><strong>Note:</strong> Newer annotations versions may be added upon request and may be already available. Please contact <a href=mailto:vishal.koparde@nih.gov>Vishal Koparde</a> for details.</p> </blockquote> <p>However, building new reference genomes is easy!</p> <p>If you do not have access to Biowulf or you are looking for a reference genome and/or annotation <strong><em>that is currently not available</em></strong>, it can be built with RENEE's build sub-command. Given a genomic FASTA file (ref.fa) and a GTF file (genes.gtf), <code>renee build</code> will create all of the required reference files to run the RENEE pipeline. Once the build pipeline completes, you can supply the newly generated reference.json to the <code>--genome</code> of <code>renee run</code>. For more information, please see the help page for the run and build sub commands.</p> <h2 id=2-tools-and-versions>2. Tools and versions<a class=headerlink href=#2-tools-and-versions title="Permanent link">&para;</a></h2> <blockquote> <p><em>Raw data &gt; Adapter Trimming &gt; Alignment &gt; Quantification (genes and isoforms, gene-fusions)</em></p> </blockquote> <table> <thead> <tr> <th><strong>Tool</strong></th> <th style="text-align: center;"><strong>Version</strong></th> <th style="text-align: left;"><strong>Docker</strong></th> <th><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>FastQC<sup>2</sup></td> <td style="text-align: center;">0.11.9</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_fastqc_0.11.9>nciccbr/ccbr_fastqc_0.11.9</a></td> <td><strong>Quality-control step</strong> to assess sequencing quality, run before and after adapter trimming</td> </tr> <tr> <td>Cutadapt<sup>3</sup></td> <td style="text-align: center;">1.18</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_cutadapt_1.18>nciccbr/ccbr_cutadapt_1.18</a></td> <td><strong>Data processing step</strong> to remove adapter sequences and perform quality trimming</td> </tr> <tr> <td>Kraken<sup>4</sup></td> <td style="text-align: center;">2.1.1</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_kraken_v2.1.1>nciccbr/ccbr_kraken_v2.1.1</a></td> <td><strong>Quality-control step</strong> to assess microbial taxonomic composition</td> </tr> <tr> <td>KronaTools<sup>5</sup></td> <td style="text-align: center;">2.7.1</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_kraken_v2.1.1>nciccbr/ccbr_kraken_v2.1.1</a></td> <td><strong>Quality-control step</strong> to visualize kraken output</td> </tr> <tr> <td>FastQ Screen<sup>6</sup></td> <td style="text-align: center;">0.13.0</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_fastq_screen_0.13.0>nciccbr/ccbr_fastq_screen_0.13.0</a></td> <td><strong>Quality-control step</strong> to assess contamination; additional dependencies: <code>bowtie2/2.3.4</code>, <code>perl/5.24.3</code></td> </tr> <tr> <td>STAR<sup>7</sup></td> <td style="text-align: center;">2.7.6a</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_arriba_2.0.0>nciccbr/ccbr_arriba_2.0.0</a></td> <td><strong>Data processing step</strong> to align reads against reference genome (using its two-pass mode)</td> </tr> <tr> <td>bbtools<sup>8</sup></td> <td style="text-align: center;">38.87</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_bbtools_38.87>nciccbr/ccbr_bbtools_38.87</a></td> <td><strong>Quality-control step</strong> to calculate insert_size of assembled reads pairs with <code>bbmerge</code></td> </tr> <tr> <td>QualiMap<sup>9</sup></td> <td style="text-align: center;">2.2.1</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_qualimap>nciccbr/ccbr_qualimap</a></td> <td><strong>Quality-control step</strong> to assess various alignment metrics</td> </tr> <tr> <td>Picard<sup>10</sup></td> <td style="text-align: center;">2.18.20</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_picard>nciccbr/ccbr_picard</a></td> <td><strong>Quality-control step</strong> to run <code>MarkDuplicates</code>, <code>CollectRnaSeqMetrics</code> and <code>AddOrReplaceReadGroups</code></td> </tr> <tr> <td>Preseq<sup>11</sup></td> <td style="text-align: center;">2.0.3</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_preseq>nciccbr/ccbr_preseq</a></td> <td><strong>Quality-control step</strong> to estimate library complexity</td> </tr> <tr> <td>SAMtools<sup>12</sup></td> <td style="text-align: center;">1.7</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_arriba_2.0.0>nciccbr/ccbr_arriba_2.0.0</a></td> <td><strong>Quality-control step</strong> to run <code>flagstat</code> to calculate alignment statistics</td> </tr> <tr> <td>bam2strandedbw</td> <td style="text-align: center;"><a href=https://github.com/CCBR/Pipeliner/blob/master/Results-template/Scripts/bam2strandedbw.pe.sh>custom</a></td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_bam2strandedbw>nciccbr/ccbr_bam2strandedbw</a></td> <td><strong>Summarization step</strong> to convert STAR aligned PE bam file into forward and reverse strand bigwigs suitable for a genomic track viewer like IGV</td> </tr> <tr> <td>RSeQC<sup>13</sup></td> <td style="text-align: center;">4.0.0</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_rseqc_4.0.0>nciccbr/ccbr_rseqc_4.0.0</a></td> <td><strong>Quality-control step</strong> to infer stranded-ness and read distributions over specific genomic features</td> </tr> <tr> <td>RSEM<sup>14</sup></td> <td style="text-align: center;">1.3.3</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_rsem_1.3.3>nciccbr/ccbr_rsem_1.3.3</a></td> <td><strong>Data processing step</strong> to quantify gene and isoform counts</td> </tr> <tr> <td>Arriba<sup>15<sup></td> <td style="text-align: center;">2.0.0</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_arriba_2.0.0>nciccbr/ccbr_arriba_2.0.0</a></td> <td><strong>Data processing step</strong> to quantify gene-fusions</td> </tr> <tr> <td>RNA Report</td> <td style="text-align: center;"><a href=https://github.com/CCBR/rNA>custom</a></td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/nciccbr/ccbr_rna>nciccbr/ccbr_rna</a></td> <td><strong>Summarization step</strong> to identify outliers and assess technical sources of variation</td> </tr> <tr> <td>MultiQC<sup>16</sup></td> <td style="text-align: center;">1.12</td> <td style="text-align: left;"><a href=https://hub.docker.com/repository/docker/skchronicles/multiqc/ >skchronicles/multiqc</a></td> <td><strong>Reporting step</strong> to aggregate sample statistics and quality-control information across all sample</td> </tr> </tbody> </table> <h2 id=3-acknowledgements>3. Acknowledgements<a class=headerlink href=#3-acknowledgements title="Permanent link">&para;</a></h2> <h3 id=31-biowulf>3.1 Biowulf<a class=headerlink href=#31-biowulf title="Permanent link">&para;</a></h3> <p>If you <a href=https://hpc.nih.gov/Research/ >utilized NIH's Biowulf cluster</a> to run RENEE, <em>please do not forget to provide an acknowlegement</em>!</p> <blockquote> <p>The continued growth and support of NIH's Biowulf cluster is dependent upon its demonstrable value to the NIH Intramural Research Program. If you publish research that involved significant use of Biowulf, please cite the cluster.</p> </blockquote> <p><strong>Suggested citation text:</strong></p> <div class=highlight><pre><span></span><code>This work utilized the computational resources of the NIH HPC Biowulf cluster. (http://hpc.nih.gov)
</code></pre></div> <h2 id=4-references>4. References<a class=headerlink href=#4-references title="Permanent link">&para;</a></h2> <p><sup><strong>1.</strong> Harrow, J., et al., GENCODE: the reference human genome annotation for The ENCODE Project. Genome Res, 2012. 22(9): p. 1760-74.</sup><br> <sup><strong>2.</strong> Andrews, S. (2010). FastQC: a quality control tool for high throughput sequence data.</sup><br> <sup><strong>3.</strong> Martin, M. (2011). "Cutadapt removes adapter sequences from high-throughput sequencing reads." EMBnet 17(1): 10-12.</sup><br> <sup><strong>4.</strong> Wood, D. E. and S. L. Salzberg (2014). "Kraken: ultrafast metagenomic sequence classification using exact alignments." Genome Biol 15(3): R46.</sup><br> <sup><strong>5.</strong> Ondov, B. D., et al. (2011). "Interactive metagenomic visualization in a Web browser." BMC Bioinformatics 12(1): 385.</sup><br> <sup><strong>6.</strong> Wingett, S. and S. Andrews (2018). "FastQ Screen: A tool for multi-genome mapping and quality control." F1000Research 7(2): 1338.</sup><br> <sup><strong>7.</strong> Dobin, A., et al., STAR: ultrafast universal RNA-seq aligner. Bioinformatics, 2013. 29(1): p. 15-21.</sup><br> <sup><strong>8.</strong> Bushnell, B., Rood, J., &amp; Singer, E. (2017). BBMerge - Accurate paired shotgun read merging via overlap. PloS one, 12(10), e0185056.</sup><br> <sup><strong>9.</strong> Okonechnikov, K., et al. (2015). "Qualimap 2: advanced multi-sample quality control for high-throughput sequencing data." Bioinformatics 32(2): 292-294.</sup><br> <sup><strong>10.</strong> The Picard toolkit. https://broadinstitute.github.io/picard/.</sup><br> <sup><strong>11.</strong> Daley, T. and A.D. Smith, Predicting the molecular complexity of sequencing libraries. Nat Methods, 2013. 10(4): p. 325-7.</sup><br> <sup><strong>12.</strong> Li, H., et al. (2009). "The Sequence Alignment/Map format and SAMtools." Bioinformatics 25(16): 2078-2079.</sup><br> <sup><strong>13.</strong> Wang, L., et al. (2012). "RSeQC: quality control of RNA-seq experiments." Bioinformatics 28(16): 2184-2185.</sup><br> <sup><strong>14.</strong> Li, B. and C.N. Dewey, RSEM: accurate transcript quantification from RNA-Seq data with or without a reference genome. BMC Bioinformatics, 2011. 12: p. 323.</sup><br> <sup><strong>15.</strong> Uhrig, S., et al. (2021). "Accurate and efficient detection of gene fusions from RNA sequencing data". Genome Res. 31(3): 448-460.</sup><br> <sup><strong>16.</strong> Ewels, P., et al. (2016). "MultiQC: summarize analysis results for multiple tools and samples in a single report." Bioinformatics 32(19): 3047-3048.</sup></p> <!---
## Future Inclusions
<sup>**NA.**    Law, C.W., et al., voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biol, 2014. 15(2): p. R29.</sup>
<sup>**NA.**    Smyth, G.K., Linear models and empirical bayes methods for assessing differential expression in microarray experiments. Stat Appl Genet Mol Biol, 2004. 3: p. Article3.</sup>
<sup>**NA.**    Fabregat, A., et al., The Reactome Pathway Knowledgebase. Nucleic Acids Res, 2018. 46(D1): p. D649-D655.</sup>
<sup>**NA.**    Liberzon, A., et al., Molecular signatures database (MSigDB) 3.0. Bioinformatics, 2011. 27(12): p. 1739-40.</sup>
<sup>**NA.**    Love, M. I., et al. (2014). "Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2." Genome Biol 15(12): 550.</sup>
<sup>**NA.**    R Core Team (2018). R: A Language and Environment for Statistical Computing. Vienna, Austria, R Foundation for Statistical Computing.</sup>
<sup>**NA.**    Leng, N., et al. (2013). "EBSeq: an empirical Bayes hierarchical model for inference in RNA-seq experiments." Bioinformatics 29(8): 1035-1043.</sup>
<sup>**NA.**    Robinson, M. D., et al. (2009). "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data." Bioinformatics 26(1): 139-140.</sup>
---> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/RENEE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/RENEE</a> </span> <span><a href=../output/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../../general-questions/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 