<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="RENEE is an open-source, reproducible, and scalable best practices pipeline for analyzing RNA-sequencing data."><meta name=author content=CCBR><link href=RNA-seq/Theory/ rel=next><link rel=icon href=assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.5.47"><title>RENEE Documentation</title><link rel=stylesheet href=assets/stylesheets/main.6f8fc17f.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#renee-rna-sequencing-analysis-pipeline class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=. title="RENEE Documentation" class="md-header__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RENEE Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> About </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class="md-tabs__item md-tabs__item--active"> <a href=. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=RNA-seq/Theory/ class=md-tabs__link> Theory </a> </li> <li class=md-tabs__item> <a href=RNA-seq/TLDR-RNA-seq/ class=md-tabs__link> Getting started </a> </li> <li class=md-tabs__item> <a href=RNA-seq/run/ class=md-tabs__link> Command Line </a> </li> <li class=md-tabs__item> <a href=RNA-seq/gui/ class=md-tabs__link> Graphical Interface </a> </li> <li class=md-tabs__item> <a href=RNA-seq/output/ class=md-tabs__link> Expected Output </a> </li> <li class=md-tabs__item> <a href=RNA-seq/Resources/ class=md-tabs__link> Resources </a> </li> <li class=md-tabs__item> <a href=general-questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=. title="RENEE Documentation" class="md-nav__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=assets/icons/doc-book.svg alt=logo> </a> RENEE Documentation </label> <div class=md-nav__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> About </span> <span class="md-nav__icon md-icon"></span> </label> <a href=. class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> About </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#table-of-contents class=md-nav__link> <span class=md-ellipsis> Table of Contents </span> </a> </li> <li class=md-nav__item> <a href=#1-introduction class=md-nav__link> <span class=md-ellipsis> 1. Introduction </span> </a> </li> <li class=md-nav__item> <a href=#2-overview class=md-nav__link> <span class=md-ellipsis> 2. Overview </span> </a> <nav class=md-nav aria-label="2. Overview"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-renee-pipeline class=md-nav__link> <span class=md-ellipsis> 2.1 RENEE Pipeline </span> </a> </li> <li class=md-nav__item> <a href=#22-reference-genomes class=md-nav__link> <span class=md-ellipsis> 2.2 Reference Genomes </span> </a> </li> <li class=md-nav__item> <a href=#23-dependencies class=md-nav__link> <span class=md-ellipsis> 2.3 Dependencies </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-run-renee-pipeline class=md-nav__link> <span class=md-ellipsis> 3. Run RENEE pipeline </span> </a> <nav class=md-nav aria-label="3. Run RENEE pipeline"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-biowulf class=md-nav__link> <span class=md-ellipsis> 3.1 Biowulf </span> </a> </li> <li class=md-nav__item> <a href=#32-frce class=md-nav__link> <span class=md-ellipsis> 3.2 FRCE </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#4-references class=md-nav__link> <span class=md-ellipsis> 4. References </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=RNA-seq/Theory/ class=md-nav__link> <span class=md-ellipsis> Theory </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/TLDR-RNA-seq/ class=md-nav__link> <span class=md-ellipsis> Getting started </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Command Line </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Command Line </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=RNA-seq/run/ class=md-nav__link> <span class=md-ellipsis> run </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/build/ class=md-nav__link> <span class=md-ellipsis> build </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/cache/ class=md-nav__link> <span class=md-ellipsis> cache </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/unlock/ class=md-nav__link> <span class=md-ellipsis> unlock </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=RNA-seq/gui/ class=md-nav__link> <span class=md-ellipsis> Graphical Interface </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/output/ class=md-nav__link> <span class=md-ellipsis> Expected Output </span> </a> </li> <li class=md-nav__item> <a href=RNA-seq/Resources/ class=md-nav__link> <span class=md-ellipsis> Resources </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=general-questions/ class=md-nav__link> <span class=md-ellipsis> General Questions </span> </a> </li> <li class=md-nav__item> <a href=troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> <li class=md-nav__item> <a href=contributing/ class=md-nav__link> <span class=md-ellipsis> How to contribute </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <!-- the README is inserted here automatically--> <h1 id=renee-rna-sequencing-analysis-pipeline>RENEE - <strong>R</strong>na s<strong>E</strong>quencing a<strong>N</strong>alysis pip<strong>E</strong>lin<strong>E</strong><a class=headerlink href=#renee-rna-sequencing-analysis-pipeline title="Permanent link">&para;</a></h1> <p>An open-source, reproducible, and scalable solution for analyzing RNA-seq data.</p> <p><a href=https://github.com/CCBR/RENEE/actions/workflows/build.yml><img alt=build src=https://github.com/CCBR/RENEE/actions/workflows/build.yml/badge.svg></a> <a href=https://github.com/CCBR/RENEE/actions/workflows/docs-mkdocs.yml><img alt=docs src=https://github.com/CCBR/RENEE/actions/workflows/docs-mkdocs.yml/badge.svg></a> <a href=https://zenodo.org/doi/10.5281/zenodo.10553198><img alt=DOI src=https://zenodo.org/badge/447297455.svg></a></p> <p>See the website for detailed information, documentation, and examples: <a href=https://ccbr.github.io/RENEE/latest/ >https://ccbr.github.io/RENEE/latest/</a></p> <h3 id=table-of-contents>Table of Contents<a class=headerlink href=#table-of-contents title="Permanent link">&para;</a></h3> <ul> <li><a href=#renee---rna-sequencing-analysis-pipeline>RENEE - <strong>R</strong>na s<strong>E</strong>quencing a<strong>N</strong>alysis pip<strong>E</strong>lin<strong>E</strong></a></li> <li><a href=#table-of-contents>Table of Contents</a></li> <li><a href=#1-introduction>1. Introduction</a></li> <li><a href=#2-overview>2. Overview</a><ul> <li><a href=#21-renee-pipeline>2.1 RENEE Pipeline</a></li> <li><a href=#22-reference-genomes>2.2 Reference Genomes</a></li> <li><a href=#23-dependencies>2.3 Dependencies</a></li> </ul> </li> <li><a href=#3-run-renee-pipeline>3. Run RENEE pipeline</a><ul> <li><a href=#31-biowulf>3.1 Biowulf</a></li> <li><a href=#32-frce>3.2 FRCE</a></li> </ul> </li> <li><a href=#4-references>4. References</a></li> <li><a href=https://github.com/CCBR/RENEE/blob/main/CHANGELOG.md>5. Version Notes</a></li> </ul> <h3 id=1-introduction>1. Introduction<a class=headerlink href=#1-introduction title="Permanent link">&para;</a></h3> <p>RNA-sequencing (<em>RNA-seq</em>) has a wide variety of applications. This popular transcriptome profiling technique can be used to quantify gene and isoform expression, detect alternative splicing events, predict gene-fusions, call variants and much more.</p> <p><strong>RENEE</strong> is a comprehensive, open-source RNA-seq pipeline that relies on technologies like <a href=https://www.docker.com/why-docker>Docker<sup>20</sup></a> and <a href=https://apptainer.org/docs/ >Singularity<sup>21</sup>... now called Apptainer</a> to maintain the highest-level of reproducibility. The pipeline consists of a series of data processing and quality-control steps orchestrated by <a href=https://snakemake.readthedocs.io/en/stable/ >Snakemake<sup>19</sup></a>, a flexible and scalable workflow management system, to submit jobs to a cluster or cloud provider.</p> <p><img alt=RENEE_overview_diagram src=https://raw.githubusercontent.com/CCBR/RENEE/5e88c208846d8098a0e26570243cb987bc2d7402/resources/overview.svg><br> <sup><strong>Fig 1. Run locally on a compute instance, on-premise using a cluster, or on the cloud using AWS.</strong> A user can define the method or mode of execution. The pipeline can submit jobs to a cluster using a job scheduler like SLURM, or run on AWS using Tibanna (feature coming soon!). A hybrid approach ensures the pipeline is accessible to all users. As an optional step, relevelant output files and metadata can be stored in object storage using HPC DME (NIH users) or Amazon S3 for archival purposes (coming soon!).</sup></p> <h3 id=2-overview>2. Overview<a class=headerlink href=#2-overview title="Permanent link">&para;</a></h3> <h4 id=21-renee-pipeline>2.1 RENEE Pipeline<a class=headerlink href=#21-renee-pipeline title="Permanent link">&para;</a></h4> <p>A bioinformatics pipeline is more than the sum of its data processing steps. A pipeline without quality-control steps provides a myopic view of the potential sources of variation within your data (i.e., biological verses technical sources of variation). RENEE pipeline is composed of a series of quality-control and data processing steps.</p> <p>The accuracy of the downstream interpretations made from transcriptomic data are highly dependent on initial sample library. Unwanted sources of technical variation, which if not accounted for properly, can influence the results. RENEE's comprehensive quality-control helps ensure your results are reliable and <em>reproducible across experiments</em>. In the data processing steps, RENEE quantifies gene and isoform expression and predicts gene fusions. Please note that the detection of alternative splicing events and variant calling will be incorporated in a later release.</p> <p><img alt="RNA-seq quantification pipeline" src=https://raw.githubusercontent.com/CCBR/RENEE/5e88c208846d8098a0e26570243cb987bc2d7402/resources/RENEE_Pipeline.svg> <sup><strong>Fig 2. An Overview of RENEE Pipeline.</strong> Gene and isoform counts are quantified and a series of QC-checks are performed to assess the quality of the data. This pipeline stops at the generation of a raw counts matrix and gene-fusion calling. To run the pipeline, a user must select their raw data, a reference genome, and output directory (i.e., the location where the pipeline performs the analysis). Quality-control information is summarized across all samples in a MultiQC report.</sup></p> <p><strong>Quality Control</strong><br> <a href=https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ ><em>FastQC</em><sup>2</sup></a> is used to assess the sequencing quality. FastQC is run twice, before and after adapter trimming. It generates a set of basic statistics to identify problems that can arise during sequencing or library preparation. FastQC will summarize per base and per read QC metrics such as quality scores and GC content. It will also summarize the distribution of sequence lengths and will report the presence of adapter sequences.</p> <p><a href=http://ccb.jhu.edu/software/kraken2/ ><em>Kraken2</em><sup>14</sup></a> and <a href=https://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/ ><em>FastQ Screen</em><sup>17</sup></a> are used to screen for various sources of contamination. During the process of sample collection to library preparation, there is a risk for introducing wanted sources of DNA. FastQ Screen compares your sequencing data to a set of different reference genomes to determine if there is contamination. It allows a user to see if the composition of your library matches what you expect. Also, if there are high levels of microbial contamination, Kraken can provide an estimation of the taxonomic composition. Kraken can be used in conjunction with <a href=https://github.com/marbl/Krona/wiki/KronaTools><em>Krona</em><sup>15</sup></a> to produce interactive reports.</p> <p><a href=http://smithlabresearch.org/software/preseq/ ><em>Preseq</em><sup>1</sup></a> is used to estimate the complexity of a library for each samples. If the duplication rate is very high, the overall library complexity will be low. Low library complexity could signal an issue with library preparation where very little input RNA was over-amplified or the sample may be degraded.</p> <p><a href=https://broadinstitute.github.io/picard/ ><em>Picard</em><sup>10</sup></a> can be used to estimate the duplication rate, and it has another particularly useful sub-command called CollectRNAseqMetrics which reports the number and percentage of reads that align to various regions: such as coding, intronic, UTR, intergenic and ribosomal regions. This is particularly useful as you would expect a library constructed with ploy(A)-selection to have a high percentage of reads that map to coding regions. Picard CollectRNAseqMetrics will also report the uniformity of coverage across all genes, which is useful for determining whether a sample has a 3' bias (observed in ploy(A)-selection libraries containing degraded RNA).</p> <p><a href=http://rseqc.sourceforge.net/ ><em>RSeQC</em><sup>9</sup></a> is another particularity useful package that is tailored for RNA-seq data. It is used to calculate the inner distance between paired-end reads and calculate TIN values for a set of canonical protein-coding transcripts. A median TIN value is calucated for each sample, which analogous to a computationally derived RIN.</p> <p><a href=https://multiqc.info/ >MultiQC<sup>11</sup></a> is used to aggregate the results of each tool into a single interactive report.</p> <p><strong>Quantification</strong><br> <a href=https://cutadapt.readthedocs.io/en/stable/ ><em>Cutadapt</em><sup>3</sup></a> is used to remove adapter sequences, perform quality trimming, and remove very short sequences that would otherwise multi-map all over the genome prior to alignment.</p> <p><a href=https://github.com/alexdobin/STAR><em>STAR</em><sup>4</sup></a> is used to align reads to the reference genome. The RENEE pipeline runs STAR in a two-passes where splice-junctions are collected and aggregated across all samples and provided to the second-pass of STAR. In the second pass of STAR, the splice-junctions detected in the first pass are inserted into the genome indices prior to alignment.</p> <p><a href=https://github.com/deweylab/RSEM><em>RSEM</em><sup>5</sup></a> is used to quantify gene and isoform expression. The expected counts from RSEM are merged across samples to create a two counts matrices for gene counts and isoform counts.</p> <p><a href=https://arriba.readthedocs.io/en/latest/ ><em>Arriba</em><sup>22</sup></a> is used to predict gene-fusion events. The pre-built human and mouse reference genomes use Arriba blacklists to reduce the false-positive rate.</p> <h4 id=22-reference-genomes>2.2 Reference Genomes<a class=headerlink href=#22-reference-genomes title="Permanent link">&para;</a></h4> <p>Pre-built reference genomes are provided on Biowulf and FRCE for a number of different annotation versions, view the list here: <a href=https://ccbr.github.io/RENEE/latest/RNA-seq/Resources/#1-reference-genomes>https://ccbr.github.io/RENEE/latest/RNA-seq/Resources/#1-reference-genomes</a></p> <p>If you would like to use a custom reference that is not already listed above, you can prepare it with the <code>renee build</code> command. See docs here: <a href=https://ccbr.github.io/RENEE/latest/RNA-seq/build/ >https://ccbr.github.io/RENEE/latest/RNA-seq/build/</a></p> <h4 id=23-dependencies>2.3 Dependencies<a class=headerlink href=#23-dependencies title="Permanent link">&para;</a></h4> <p><strong>Requires:</strong> <code>singularity&gt;=3.5</code> <code>snakemake&gt;=6.0</code></p> <blockquote> <p><strong>NOTE:</strong><br> <ins>Biowulf users</ins>:<br> Both, singularity and snakemake, modules are already installed and available for all Biowulf users. Please skip this step as <code>module load ccbrpipeliner</code> will preload singularity and snakemake.</p> </blockquote> <p><a href=https://snakemake.readthedocs.io/en/stable/getting_started/installation.html>Snakemake</a> and <a href=https://singularity.lbl.gov/all-releases>singularity</a> must be installed on the target system. Snakemake orchestrates the execution of each step in the pipeline. To guarantee reproducibility, each step relies on pre-built images from <a href=https://hub.docker.com/orgs/nciccbr/repositories>DockerHub</a>. Snakemake pulls these docker images while converting them to singularity on the fly and saves them onto the local filesystem prior to job execution, and as so, snakemake and singularity are the only two dependencies.</p> <hr> <p align=center> <a href=#renee---rna-sequencing-analysis-pipeline>Back to Top</a> </p> <hr> <h3 id=3-run-renee-pipeline>3. Run RENEE pipeline<a class=headerlink href=#3-run-renee-pipeline title="Permanent link">&para;</a></h3> <h4 id=31-biowulf>3.1 Biowulf<a class=headerlink href=#31-biowulf title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=c1># RENEE is configured to use different execution backends: local or slurm</span>
<span class=c1># view the help page for more information</span>
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
renee<span class=w> </span>run<span class=w> </span>--help

<span class=c1># @local: uses local singularity execution method</span>
<span class=c1># The local MODE will run serially on compute</span>
<span class=c1># instance. This is useful for testing, debugging,</span>
<span class=c1># or when a users does not have access to a high</span>
<span class=c1># performance computing environment.</span>
<span class=c1># Please note that you can dry-run the command below</span>
<span class=c1># by providing the --dry-run flag</span>
<span class=c1># Do not run this on the head node!</span>
<span class=c1># Grab an interactive node</span>
sinteractive<span class=w> </span>--mem<span class=o>=</span>110g<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>12</span><span class=w> </span>--gres<span class=o>=</span>lscratch:200
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
renee<span class=w> </span>run<span class=w> </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/RNA_hg38<span class=w> </span>--genome<span class=w> </span>hg38_36<span class=w> </span>--mode<span class=w> </span><span class=nb>local</span>

<span class=c1># @slurm: uses slurm and singularity execution method</span>
<span class=c1># The slurm MODE will submit jobs to the cluster.</span>
<span class=c1># The --sif-cache flag will re-use singularity containers from a shared location.</span>
<span class=c1># It is recommended running RENEE in this mode.</span>
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
renee<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/RNA_hg38<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--genome<span class=w> </span>hg38_36<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--sif-cache<span class=w> </span>/data/CCBR_Pipeliner/SIFs
</code></pre></div> <h4 id=32-frce>3.2 FRCE<a class=headerlink href=#32-frce title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=c1># grab an interactive node</span>
srun<span class=w> </span>--export<span class=w> </span>all<span class=w> </span>--pty<span class=w> </span>--x11<span class=w> </span>bash

<span class=c1># add renee to path correctly</span>
.<span class=w> </span>/mnt/projects/CCBR-Pipelines/pipelines/guis/latest/bin/setup

<span class=c1># run renee</span>
renee<span class=w> </span>--help
</code></pre></div> <p>When running renee on FRCE, we recommend setting <code>--tmp-dir</code> and <code>--sif-cache</code> with the following values:</p> <div class=highlight><pre><span></span><code>renee<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--input<span class=w> </span>.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--output<span class=w> </span>/scratch/cluster_scratch/<span class=nv>$USER</span>/RNA_hg38<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--genome<span class=w> </span>hg38_36<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--tmp-dir<span class=w> </span>/scratch/cluster_scratch/<span class=nv>$USER</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--sif-cache<span class=w> </span>/mnt/projects/CCBR-Pipelines/SIFs
</code></pre></div> <hr> <p align=center> <a href=#renee---rna-sequencing-analysis-pipeline>Back to Top</a> </p> <hr> <h3 id=4-references>4. References<a class=headerlink href=#4-references title="Permanent link">&para;</a></h3> <p><sup><strong>1.</strong> Daley, T. and A.D. Smith, Predicting the molecular complexity of sequencing libraries. Nat Methods, 2013. 10(4): p. 325-7.</sup><br> <sup><strong>2.</strong> Andrews, S. (2010). FastQC: a quality control tool for high throughput sequence data.</sup><br> <sup><strong>3.</strong> Martin, M. (2011). "Cutadapt removes adapter sequences from high-throughput sequencing reads." EMBnet 17(1): 10-12.</sup><br> <sup><strong>4.</strong> Dobin, A., et al., STAR: ultrafast universal RNA-seq aligner. Bioinformatics, 2013. 29(1): p. 15-21.</sup><br> <sup><strong>5.</strong> Li, B. and C.N. Dewey, RSEM: accurate transcript quantification from RNA-Seq data with or without a reference genome. BMC Bioinformatics, 2011. 12: p. 323.</sup><br> <sup><strong>6.</strong> Harrow, J., et al., GENCODE: the reference human genome annotation for The ENCODE Project. Genome Res, 2012. 22(9): p. 1760-74.</sup><br> <sup><strong>7.</strong> Law, C.W., et al., voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biol, 2014. 15(2): p. R29.</sup><br> <sup><strong>8.</strong> Smyth, G.K., Linear models and empirical bayes methods for assessing differential expression in microarray experiments. Stat Appl Genet Mol Biol, 2004. 3: p. Article3.</sup><br> <sup><strong>9.</strong> Wang, L., et al. (2012). "RSeQC: quality control of RNA-seq experiments." Bioinformatics 28(16): 2184-2185.</sup><br> <sup><strong>10.</strong> The Picard toolkit. <a href=https://broadinstitute.github.io/picard/ >https://broadinstitute.github.io/picard/</a>.</sup><br> <sup><strong>11.</strong> Ewels, P., et al. (2016). "MultiQC: summarize analysis results for multiple tools and samples in a single report." Bioinformatics 32(19): 3047-3048.</sup><br> <sup><strong>12.</strong> R Core Team (2018). R: A Language and Environment for Statistical Computing. Vienna, Austria, R Foundation for Statistical Computing.</sup><br> <sup><strong>13.</strong> Li, H., et al. (2009). "The Sequence Alignment/Map format and SAMtools." Bioinformatics 25(16): 2078-2079.</sup><br> <sup><strong>14.</strong> Wood, D. E. and S. L. Salzberg (2014). "Kraken: ultrafast metagenomic sequence classification using exact alignments." Genome Biol 15(3): R46.</sup><br> <sup><strong>15.</strong> Ondov, B. D., et al. (2011). "Interactive metagenomic visualization in a Web browser." BMC Bioinformatics 12(1): 385.</sup><br> <sup><strong>16.</strong> Okonechnikov, K., et al. (2015). "Qualimap 2: advanced multi-sample quality control for high-throughput sequencing data." Bioinformatics 32(2): 292-294.</sup><br> <sup><strong>17.</strong> Wingett, S. and S. Andrews (2018). "FastQ Screen: A tool for multi-genome mapping and quality control." F1000Research 7(2): 1338.</sup><br> <sup><strong>18.</strong> Robinson, M. D., et al. (2009). "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data." Bioinformatics 26(1): 139-140.</sup><br> <sup><strong>19.</strong> Koster, J. and S. Rahmann (2018). "Snakemake-a scalable bioinformatics workflow engine." Bioinformatics 34(20): 3600.</sup><br> <sup><strong>20.</strong> Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2.</sup><br> <sup><strong>21.</strong> Kurtzer GM, Sochat V, Bauer MW (2017). Singularity: Scientific containers for mobility of compute. PLoS ONE 12(5): e0177459.</sup><br> <sup><strong>22.</strong> Haas, B. J., et al. (2019). "Accuracy assessment of fusion transcript detection via read-mapping and de novo fusion transcript assembly-based methods." Genome Biology 20(1): 213.</sup></p> <hr> <p align=center> <a href=#renee---rna-sequencing-analysis-pipeline>Back to Top</a> </p> <hr> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> 2024-08-16 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2023 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160m368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160M0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7M405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.4zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0m-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.3c73.7 0 133.4 59.7 133.4 133.3 0 14.7-11.9 26.7-26.7 26.7H154.6c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=assets/javascripts/bundle.83f73b43.min.js></script> </body> </html>