<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="RENEE is an open-source, reproducible, and scalable best practices pipeline for analyzing RNA-sequencing data."><meta name=author content=CCBR><link href=../run/ rel=prev><link href=../cache/ rel=next><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.5.46"><title>build - RENEE Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.6f8fc17f.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#renee-build class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="RENEE Documentation" class="md-header__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RENEE Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> build </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../Theory/ class=md-tabs__link> Theory </a> </li> <li class=md-tabs__item> <a href=../TLDR-RNA-seq/ class=md-tabs__link> Getting started </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../run/ class=md-tabs__link> Command Line </a> </li> <li class=md-tabs__item> <a href=../gui/ class=md-tabs__link> Graphical Interface </a> </li> <li class=md-tabs__item> <a href=../output/ class=md-tabs__link> Expected Output </a> </li> <li class=md-tabs__item> <a href=../Resources/ class=md-tabs__link> Resources </a> </li> <li class=md-tabs__item> <a href=../../general-questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="RENEE Documentation" class="md-nav__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> RENEE Documentation </label> <div class=md-nav__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> About </span> </a> </li> <li class=md-nav__item> <a href=../Theory/ class=md-nav__link> <span class=md-ellipsis> Theory </span> </a> </li> <li class=md-nav__item> <a href=../TLDR-RNA-seq/ class=md-nav__link> <span class=md-ellipsis> Getting started </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Command Line </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Command Line </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../run/ class=md-nav__link> <span class=md-ellipsis> run </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> build </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> build </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-about class=md-nav__link> <span class=md-ellipsis> 1. About </span> </a> </li> <li class=md-nav__item> <a href=#2-synopsis class=md-nav__link> <span class=md-ellipsis> 2. Synopsis </span> </a> <nav class=md-nav aria-label="2. Synopsis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-required-arguments class=md-nav__link> <span class=md-ellipsis> 2.1 Required Arguments </span> </a> </li> <li class=md-nav__item> <a href=#22-build-options class=md-nav__link> <span class=md-ellipsis> 2.2 Build Options </span> </a> </li> <li class=md-nav__item> <a href=#23-orchestration-options class=md-nav__link> <span class=md-ellipsis> 2.3 Orchestration Options </span> </a> </li> <li class=md-nav__item> <a href=#24-misc-options class=md-nav__link> <span class=md-ellipsis> 2.4 Misc Options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#3-hybrid-genomes class=md-nav__link> <span class=md-ellipsis> 3. Hybrid Genomes </span> </a> </li> <li class=md-nav__item> <a href=#4-convert-ncbi-gff3-to-gtf-format class=md-nav__link> <span class=md-ellipsis> 4. Convert NCBI GFF3 to GTF format </span> </a> </li> <li class=md-nav__item> <a href=#5-example class=md-nav__link> <span class=md-ellipsis> 5. Example </span> </a> <nav class=md-nav aria-label="5. Example"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#51-biowulf class=md-nav__link> <span class=md-ellipsis> 5.1 Biowulf </span> </a> </li> <li class=md-nav__item> <a href=#52-generic-slurm-cluster class=md-nav__link> <span class=md-ellipsis> 5.2 Generic SLURM Cluster </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> <span class=md-ellipsis> cache </span> </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> <span class=md-ellipsis> unlock </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../gui/ class=md-nav__link> <span class=md-ellipsis> Graphical Interface </span> </a> </li> <li class=md-nav__item> <a href=../output/ class=md-nav__link> <span class=md-ellipsis> Expected Output </span> </a> </li> <li class=md-nav__item> <a href=../Resources/ class=md-nav__link> <span class=md-ellipsis> Resources </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../general-questions/ class=md-nav__link> <span class=md-ellipsis> General Questions </span> </a> </li> <li class=md-nav__item> <a href=../../troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/ class=md-nav__link> <span class=md-ellipsis> How to contribute </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=renee-build><code>renee <b>build</b></code><a class=headerlink href=#renee-build title="Permanent link">&para;</a></h1> <h2 id=1-about>1. About<a class=headerlink href=#1-about title="Permanent link">&para;</a></h2> <p>The <code>renee</code> executable is composed of several inter-related sub commands. Please see <code>renee -h</code> for all available options.</p> <p>This part of the documentation describes options and concepts for <code>renee <b>build</b></code> sub command in more detail. With minimal configuration, the <strong><code>build</code></strong> sub command enables you to build new reference files for the renee run pipeline.</p> <p>Setting up the RENEE build pipeline is fast and easy! In its most basic form, <code>renee <b>build</b></code> only has <em>five required inputs</em>.</p> <h2 id=2-synopsis>2. Synopsis<a class=headerlink href=#2-synopsis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>$ renee build [--help] \
             [--shared-resources SHARED_RESOURCES] [--small-genome] \
             [--dry-run] [--singularity-cache SINGULARITY_CACHE] \
             [--sif-cache SIF_CACHE] [--tmp-dir TMP_DIR] \
             --ref-fa REF_FA \
             --ref-name REF_NAME \
             --ref-gtf REF_GTF \
             --gtf-ver GTF_VER \
             --output OUTPUT
</code></pre></div> <p>The synopsis for each command shows its parameters and their usage. Optional parameters are shown in square brackets.</p> <p>A user <strong>must</strong> provide the genomic sequence of the reference's assembly in FASTA format via <code>--ref-fa</code> argument, an alias for the reference genome via <code>--ref-name</code> argument, a gene annotation for the reference assembly via <code>--ref-gtf</code> argument, an alias or version for the gene annotation via the <code>--gtf-ver</code> argument, and an output directory to store the built reference files via <code>--output</code> argument. If you are running the pipeline outside of Biowulf, you will need to additionally provide the the following options: <code>--shared-resources</code>, <code>--tmp-dir</code>. More information about each of these options can be found below.</p> <p>For <a href=https://www.gencodegenes.org/human/ >human</a> and <a href=https://www.gencodegenes.org/mouse/ >mouse</a> data, we highly recommend downloading the latest available <strong>PRI</strong> genome assembly and corresponding gene annotation from <a href=https://www.gencodegenes.org/ >GENCODE</a>. These reference files contain chromosomes and scaffolds sequences.</p> <p>The build pipeline will generate a JSON file containing key, value pairs to required reference files for the <code>renee <b>run</b></code> pipeline. This file will be located in the path provided to <code>--output</code>. The name of this JSON file is dependent on the values provided to <code>--ref-name</code> and <code>--gtf-ver</code> and has the following naming convention: <code>{OUTPUT}/{REF_NAME}_{GTF_VER}.json</code>. Once the build pipeline completes, this reference JSON file can be passed to the <code>--genome</code> option of <code>renee <b>run</b></code>. This is how new references are built for the RENEE pipeline.</p> <p>Use you can always use the <code>-h</code> option for information on a specific command.</p> <h3 id=21-required-arguments>2.1 Required Arguments<a class=headerlink href=#21-required-arguments title="Permanent link">&para;</a></h3> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--ref-fa REF_FA</code></p> <blockquote> <p><strong>Genomic FASTA file of the reference genome.</strong><br> <em>type: file</em></p> <p>This file represents the genome sequence of the reference assembly in FASTA format. If you are downloading this from GENCODE, you should select the <em>PRI</em> genomic FASTA file. This file will contain the primary genomic assembly (contains chromosomes and scaffolds). This input file should not be compressed. Sequence identifiers in this file must match with sequence identifiers in the GTF file provided to <code>--ref-gtf</code>.</p> <p><strong><em>Example:</em></strong> &gt; <code>--ref-fa GRCh38.primary_assembly.genome.fa</code></p> </blockquote> <hr> <p><code>--ref-name REF_NAME</code></p> <blockquote> <p><strong>Name of the reference genome.</strong><br> <em>type: string</em></p> <p>Name or alias for the reference genome. This can be the common name for the reference genome. Here is a list of common examples for different model organisms: mm10, hg38, rn6, danRer11, dm6, canFam3, sacCer3, ce11. If the provided values contains one of the following sub-strings (hg19, hs37d, grch37, hg38, hs38d, grch38, mm10, grcm38), then Arriba will run with its corresponding blacklist.</p> <p><strong><em>Example:</em></strong> <code>--ref-name hg38</code></p> </blockquote> <hr> <p><code>--ref-gtf REF_GTF</code></p> <blockquote> <p><strong>Gene annotation or GTF file for the reference genome.</strong><br> <em>type: file</em></p> <p>This file represents the reference genome's gene annotation in GTF format. If you are downloading this from GENCODE, you should select the 'PRI' GTF file. This file contains gene annotations for the primary assembly (contains chromosomes and scaffolds). This input file should not be compressed. Sequence identifiers (column 1) in this file must match with sequence identifiers in the FASTA file provided to <code>--ref-fa</code>.<br> <strong><em>Example:</em></strong> <code>--ref-gtf gencode.v36.primary_assembly.annotation.gtf</code></p> </blockquote> <hr> <p><code>--gtf-ver GTF_VER</code></p> <blockquote> <p><strong>Version of the gene annotation or GTF file provided.</strong><br> <em>type: string or int</em></p> <p>This is the version of the supplied gene annotation or GTF file. If you are using a GTF file from GENCODE, use the release number or version (i.e. <em>M25</em> for mouse or <em>37</em> for human). Visit gencodegenes.org for more details.<br> <strong><em>Example:</em></strong> <code>--gtf-ver 36</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong><br> <em>type: path</em></p> <p>This location is where the build pipeline will create all of its output files. If the user-provided working directory has not been initialized, it will automatically be created. Note: by default, any files in <code>config</code>, <code>resources,</code> or <code>workflow</code> in the output directory may be overwritten by <code>renee build</code>. <strong><em>Example:</em></strong> <code>--output /data/$USER/refs/hg38_36/</code></p> </blockquote> <h3 id=22-build-options>2.2 Build Options<a class=headerlink href=#22-build-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional and do not need to be provided. If you are running the pipeline outside of Biowulf, the <code>--shared-resources</code> option only needs to be provided at least once. This will ensure reference files that are shared across different genomes are downloaded locally.</p> <p><code>--shared-resources SHARED_RESOURCES</code></p> <blockquote> <p><strong>Local path to shared resources.</strong><br> <em>type: path</em></p> <p>The pipeline uses a set of shared reference files that can be re-used across reference genomes. These currently include reference files for kraken and FQScreen. These reference files can be downloaded with the build sub command's <code>--shared-resources</code> option. With that being said, these files only need to be downloaded once. We recommend storing this files in a shared location on the filesystem that other people can access. If you are running the pipeline on Biowulf, you do NOT need to download these reference files! They already exist on the filesystem in a location that anyone can access; however, if you are running the pipeline on another cluster or target system, you will need to download the shared resources with the build sub command, and you will need to provide this option every time you run the pipeline. Please provide the same path that was provided to the build sub command's --shared-resources option. Again, if you are running the pipeline on Biowulf, you do NOT need to provide this option. For more information about how to download shared resources, please reference the build sub command's <code>--shared-resources</code> option.</p> <p><strong><em>Example:</em></strong> <code>--shared-resources /data/shared/renee</code></p> </blockquote> <hr> <p><code>--small-genome</code></p> <blockquote> <p><strong>Builds a small genome index.</strong><br> <em>type: boolean</em></p> <p>For small genomes, it is recommended running STAR with a scaled down <code>--genomeSAindexNbases</code> value. This option runs the build pipeline in a mode where it dynamically finds the optimal value for this option using the following formula: <code>min(14, log2(GenomeSize)/2 - 1)</code>. Generally speaking, this option is not really applicable for most mammalian reference genomes, i.e. human and mouse; however, researcher working with very small reference genomes, like S. cerevisiae ~ 12Mb, should provide this option.</p> <p>When in doubt feel free to provide this option, as the optimal value will be found based on your input.</p> <p><strong><em>Example:</em></strong> <code>--small-genome</code></p> </blockquote> <h3 id=23-orchestration-options>2.3 Orchestration Options<a class=headerlink href=#23-orchestration-options title="Permanent link">&para;</a></h3> <p><code>--dry-run</code></p> <blockquote> <p><strong>Dry run the build pipeline.</strong><br> <em>type: boolean</em></p> <p>Displays what steps in the build pipeline remain or will be run. Does not execute anything!</p> <p><strong><em>Example:</em></strong> <code>--dry-run</code></p> </blockquote> <hr> <p><code>--singularity-cache SINGULARITY_CACHE</code></p> <blockquote> <p><strong>Overrides the $SINGULARITY_CACHEDIR environment variable.</strong><br> <em>type: path</em><br> <em>default: <code>--output OUTPUT/.singularity</code></em></p> <p>Singularity will cache image layers pulled from remote registries. This ultimately speeds up the process of pull an image from DockerHub if an image layer already exists in the singularity cache directory. By default, the cache is set to the value provided to the <code>--output</code> argument. Please note that this cache cannot be shared across users. Singularity strictly enforces you own the cache directory and will return a non-zero exit code if you do not own the cache directory! See the <code>--sif-cache</code> option to create a shareable resource.</p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/.singularity</code></p> </blockquote> <hr> <p><code>--sif-cache SIF_CACHE</code></p> <blockquote> <p><strong>Path where a local cache of SIFs are stored.</strong><br> <em>type: path</em></p> <p>Uses a local cache of SIFs on the filesystem. This SIF cache can be shared across users if permissions are set correctly. If a SIF does not exist in the SIF cache, the image will be pulled from Dockerhub and a warning message will be displayed. The <code>renee cache</code> subcommand can be used to create a local SIF cache. Please see <code>renee cache</code> for more information. This command is extremely useful for avoiding DockerHub pull rate limits. It also remove any potential errors that could occur due to network issues or DockerHub being temporarily unavailable. We recommend running RENEE with this option when ever possible.</p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/SIFs</code></p> </blockquote> <hr> <p><code>--tmp-dir TMP_DIR</code></p> <blockquote> <p><strong>Path on the file system for writing temporary files.</strong><br> <em>type: path</em><br> <em>default: <code>/lscratch/$SLURM_JOBID</code></em></p> <p>Path on the file system for writing temporary output files. By default, the temporary directory is set to '/lscratch/$SLURM_JOBID' on NIH's Biowulf cluster and 'OUTPUT' on the FRCE cluster. However, if you are running the pipeline on another cluster, this option will need to be specified. Ideally, this path should point to a dedicated location on the filesystem for writing tmp files. On many systems, this location is set to somewhere in /scratch. If you need to inject avariable into this string that should NOT be expanded,please quote this options value in single quotes.</p> <p><strong><em>Example:</em></strong> <code>--tmp-dir /cluster_scratch/$USER/</code></p> </blockquote> <h3 id=24-misc-options>2.4 Misc Options<a class=headerlink href=#24-misc-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional and do not need to be provided.</p> <p><code>-h, --help</code></p> <blockquote> <p><strong>Display Help.</strong><br> <em>type: boolean</em></p> <p>Shows command's synopsis, help message, and an example command</p> <p><strong><em>Example:</em></strong> <code>--help</code></p> </blockquote> <h2 id=3-hybrid-genomes>3. Hybrid Genomes<a class=headerlink href=#3-hybrid-genomes title="Permanent link">&para;</a></h2> <p>If you have two GTF files, e.g. hybrid genomes (host + virus), then you need to create one genomic FASTA file and one GTF file for the hybrid genome prior to running the <code>renee <b>build</b></code> command.</p> <p>We recommend creating an artificial chromosome for the non-host sequence. The sequence identifier in the FASTA file must match the sequence identifier in the GTF file (column 1). Generally speaking, since the host annotation is usually downloaded from Ensembl or GENCODE, it will be correctly formatted; however, that may not be the case for the non-host sequence!</p> <p>Please ensure the non-host annotation contains the following features and/or constraints:</p> <ul> <li>for a given <code>gene</code> feature</li> <li>each <code>gene</code> entry has at least one <code>transcript</code> feature</li> <li>and each <code>transcript</code> entry has at least one <code>exon</code> feature</li> <li><code>gene_id</code>, <code>gene_name</code> and <code>gene_biotype</code> are required</li> <li>for a given <code>transcipt</code> feature</li> <li>along with <code>gene_id</code>, <code>gene_name</code> and <code>gene_biotype</code> ... <code>transcript_id</code> is also required</li> <li>for a given <code>exon</code> feature</li> <li><code>gene_id</code>, <code>gene_name</code>, <code>gene_biotype</code>, <code>transcript_id</code> are required</li> </ul> <p>If not, the GTF file may need to be manually curated until these conditions are satisfied.</p> <p>Here is an example feature from a hand-curated Biotyn_probe GTF file:</p> <div class=highlight><pre><span></span><code>Biot1<span class=w>   </span>BiotynProbe<span class=w> </span>gene<span class=w>    </span><span class=m>1</span><span class=w>   </span><span class=m>21</span><span class=w>  </span><span class=m>0</span>.000000<span class=w>    </span>+<span class=w>   </span>.<span class=w>   </span>gene_id<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>gene_name<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>gene_biotype<span class=w> </span><span class=s2>&quot;biotynlated_probe_control&quot;</span><span class=p>;</span>
Biot1<span class=w>   </span>BiotynProbe<span class=w> </span>transcript<span class=w>  </span><span class=m>1</span><span class=w>   </span><span class=m>21</span><span class=w>  </span><span class=m>0</span>.000000<span class=w>    </span>+<span class=w>   </span>.<span class=w>   </span>gene_id<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>gene_name<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>gene_biotype<span class=w> </span><span class=s2>&quot;biotynlated_probe_control&quot;</span><span class=p>;</span><span class=w> </span>transcript_id<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>transcript_name<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>transcript_type<span class=w> </span><span class=s2>&quot;biotynlated_probe_control&quot;</span><span class=p>;</span>
Biot1<span class=w>   </span>BiotynProbe<span class=w> </span>exon<span class=w>    </span><span class=m>1</span><span class=w>   </span><span class=m>21</span><span class=w>  </span><span class=m>0</span>.000000<span class=w>    </span>+<span class=w>   </span>.<span class=w>   </span>gene_id<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>gene_biotype<span class=w> </span><span class=s2>&quot;biotynlated_probe_control&quot;</span><span class=p>;</span><span class=w> </span>transcript_id<span class=w> </span><span class=s2>&quot;Biot1&quot;</span><span class=p>;</span><span class=w> </span>transcript_type<span class=w> </span><span class=s2>&quot;biotynlated_probe_control&quot;</span><span class=p>;</span>
</code></pre></div> <p>In this tab-delimited example above,</p> <ul> <li><strong><em>line 1:</em></strong> the <code>gene</code> feature has 3 required attributes in column 9: <code>gene_id</code> and <code>gene_name</code> and <code>gene_biotype</code></li> <li><strong><em>line 2:</em></strong> the <code>transcript</code> entry for the above <code>gene</code> repeats the same attributes with following required fields: <code>transcript_id</code> and <code>transcript_name</code></li> <li><em>Please note:</em> <code>transcript_type</code> is <em>optional</em></li> <li><strong><em>line 3:</em></strong> the <code>exon</code> entry for the above <code>transcript</code> has 3 required attributes: <code>gene_id</code> and <code>transcript_id</code> and <code>gene_biotype</code></li> <li><em>Please note:</em> <code>transcript_type</code> is <em>optional</em></li> </ul> <p>For a given gene, the combination of the <code>gene_id</code> AND <code>gene_name</code> should form a unique string. There should be no instances where two different genes share the same <code>gene_id</code> AND <code>gene_name</code>.</p> <h2 id=4-convert-ncbi-gff3-to-gtf-format>4. Convert NCBI GFF3 to GTF format<a class=headerlink href=#4-convert-ncbi-gff3-to-gtf-format title="Permanent link">&para;</a></h2> <p>It is worth noting that RENEE comes bundled with a script to convert GFF3 files downloaded from NCBI to GTF file format. This convenience script is useful as the <code>renee build</code> sub command takes a GTF file as one of its inputs.</p> <p>Please note that this script has only been tested with GFF3 files downloaded from NCBI, and <em>it is <strong>not</strong> recommended to use with GFF3 files originating from other sources like Ensembl or GENCODE</em>. If you are selecting an annotation from Ensembl or GENCODE, please download the GTF file option.</p> <p>The only dependency of the script is the python package argparse, which comes bundled with the following python&#8532; distributions: <code>python&gt;=2.7.18</code> or <code>python&gt;=3.2</code>. If argparse is not installed, it can be downloaded with pip by running the following command:</p> <div class=highlight><pre><span></span><code>pip<span class=w> </span>install<span class=w> </span>--upgrade<span class=w> </span>pip
pip<span class=w> </span>install<span class=w> </span>argparse
</code></pre></div> <p>For more information about the script and its usage, please run:</p> <div class=highlight><pre><span></span><code>./resources/gff3togtf.py<span class=w> </span>-h
</code></pre></div> <h2 id=5-example>5. Example<a class=headerlink href=#5-example title="Permanent link">&para;</a></h2> <h3 id=51-biowulf>5.1 Biowulf<a class=headerlink href=#51-biowulf title="Permanent link">&para;</a></h3> <p>On Biowulf getting started with the pipeline is fast and easy! In this example, we build a mouse reference genome.</p> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node (do not run on head node)</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>2</span>:00:00<span class=w> </span>-p<span class=w> </span>interactive<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>4</span><span class=w> </span>--pty<span class=w> </span>bash
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner

<span class=c1># Step 1.) Dry run the Build pipeline</span>
renee<span class=w> </span>build<span class=w> </span>--ref-fa<span class=w> </span>GRCm39.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-name<span class=w> </span>mm39<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-gtf<span class=w> </span>gencode.vM26.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--gtf-ver<span class=w> </span>M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/refs/mm39_M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--sif-cache<span class=w> </span>/data/CCBR_Pipeliner/SIFs/<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--dry-run

<span class=c1># Step 2.) Build new RENEE reference files</span>
renee<span class=w> </span>build<span class=w> </span>--ref-fa<span class=w> </span>GRCm39.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-name<span class=w> </span>mm39<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-gtf<span class=w> </span>gencode.vM26.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--gtf-ver<span class=w> </span>M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/refs/mm39_M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--sif-cache<span class=w> </span>/data/CCBR_Pipeliner/SIFs/
</code></pre></div> <h3 id=52-generic-slurm-cluster>5.2 Generic SLURM Cluster<a class=headerlink href=#52-generic-slurm-cluster title="Permanent link">&para;</a></h3> <p>Running the pipeline outside of Biowulf is easy; however, there are a few extra options you must provide. Please note when running the build sub command for the first time, you will also need to provide the <code>--shared-resources</code> option. This option will download our kraken2 database and bowtie2 indices for FastQ Screen. The path provided to this option should be provided to the <code>--shared-resources</code> option of the <a href=./RNA-seq/cache/ >run</a> sub command. Next, you will also need to provide a path to write temporary output files via the <code>--tmp-dir</code> option. We also recommend providing a path to a SIF cache. You can cache software containers locally with the <a href=./RNA-seq/cache/ >cache</a> sub command.</p> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node (do not run on head node)</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>2</span>:00:00<span class=w> </span>-p<span class=w> </span>interactive<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>4</span><span class=w> </span>--pty<span class=w> </span>bash
<span class=c1># Add snakemake and singularity to $PATH,</span>
<span class=c1># This step may vary across clusters, you</span>
<span class=c1># can reach out to a sys admin if snakemake</span>
<span class=c1># and singularity are not installed.</span>
module<span class=w> </span>purge
<span class=c1># Replace the following:</span>
<span class=c1># module load ccbrpipeliner</span>
<span class=c1># with module load statements that load</span>
<span class=c1># python &gt;= 3.7,</span>
<span class=c1># snakemake, and</span>
<span class=c1># singularity</span>
<span class=c1># before running renee</span>
<span class=c1># Also, ensure that the `renee` executable is in PATH</span>

<span class=c1># Step 1.) Dry run the Build pipeline</span>
renee<span class=w> </span>build<span class=w> </span>--ref-fa<span class=w> </span>GRCm39.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-name<span class=w> </span>mm39<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-gtf<span class=w> </span>gencode.vM26.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--gtf-ver<span class=w> </span>M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/refs/mm39_M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--shared-resources<span class=w> </span>/data/shared/renee<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--tmp-dir<span class=w> </span>/cluster_scratch/<span class=nv>$USER</span>/<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--sif-cache<span class=w> </span>/data/<span class=nv>$USER</span>/cache<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--dry-run

<span class=c1># Step 2.) Build new RENEE reference files</span>
renee<span class=w> </span>build<span class=w> </span>--ref-fa<span class=w> </span>GRCm39.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-name<span class=w> </span>mm39<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--ref-gtf<span class=w> </span>gencode.vM26.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--gtf-ver<span class=w> </span>M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/refs/mm39_M26<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--shared-resources<span class=w> </span>/data/shared/renee<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--tmp-dir<span class=w> </span>/cluster_scratch/<span class=nv>$USER</span>/<span class=w> </span><span class=se>\</span>
<span class=w>              </span>--sif-cache<span class=w> </span>/data/<span class=nv>$USER</span>/cache
</code></pre></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> 2024-08-05 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2023 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160m368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160M0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7M405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.4zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0m-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.3c73.7 0 133.4 59.7 133.4 133.3 0 14.7-11.9 26.7-26.7 26.7H154.6c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.83f73b43.min.js></script> </body> </html>