<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content=CCBR><link rel="shortcut icon" href=../../img/favicon.ico><title>Getting started - RENEE Documentation</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Getting started";
        var mkdocs_page_input_path = "RNA-seq/TLDR-RNA-seq.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> RENEE Documentation </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>About</a> </li> </ul> <ul> <li class=toctree-l1><a href=../Theory/ class="reference internal">Theory</a> </li> </ul> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Getting started</a> <ul class=current> </ul> </li> </ul> <p class=caption><span class=caption-text>Command Line</span></p> <ul> <li class=toctree-l1><a href=../run/ class="reference internal">run</a> </li> <li class=toctree-l1><a href=../build/ class="reference internal">build</a> </li> <li class=toctree-l1><a href=../cache/ class="reference internal">cache</a> </li> <li class=toctree-l1><a href=../unlock/ class="reference internal">unlock</a> </li> </ul> <ul> <li class=toctree-l1><a href=../gui/ class="reference internal">Graphical Interface</a> </li> </ul> <ul> <li class=toctree-l1><a href=../output/ class="reference internal">Expected Output</a> </li> </ul> <ul> <li class=toctree-l1><a href=../Resources/ class="reference internal">Resources</a> </li> </ul> <p class=caption><span class=caption-text>FAQ</span></p> <ul> <li class=toctree-l1><a href=../../general-questions/ class="reference internal">General Questions</a> </li> <li class=toctree-l1><a href=../../troubleshooting/ class="reference internal">Troubleshooting</a> </li> <li class=toctree-l1><a href=../../contributing/ class="reference internal">How to contribute</a> </li> </ul> <ul> <li class=toctree-l1><a href=../../license/ class="reference internal">License</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>RENEE Documentation</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Getting started</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/RENEE/edit/master/docs/RNA-seq/TLDR-RNA-seq.md>Edit on CCBR/RENEE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h2 id=1-introduction>1. Introduction<a class=headerlink href=#1-introduction title="Permanent link">&para;</a></h2> <p>When processing RNA-sequencing data, there are often many steps that we must repeat. These are usually steps like removing adapter sequences, aligning reads against a reference genome, checking the quality of the data, and quantifying counts. RENEE is composed of several sub commands or convenience functions to automate these repetitive steps.</p> <p>With RENEE, you can run your samples through our highly-reproducible pipeline, build resources for new reference genomes, and more!</p> <p>Here is a list of available renee <code>sub commands</code>:</p> <ul> <li><a href=../run><strong><code>run</code></strong></a>: run the rna-seq pipeline</li> <li><a href=../build><strong><code>build</code></strong></a>: build reference files</li> <li><a href=../cache><strong><code>cache</code></strong></a>: cache remote resources locally</li> <li><a href=../unlock><strong><code>unlock</code></strong></a>: unlock a working directory</li> </ul> <blockquote> <p>This page contains information for building reference files and running the RENEE pipeline. For more information about each of the available sub commands, please see the <a href=../run/ >usage section</a>.</p> </blockquote> <h2 id=2-setup-renee>2. Setup RENEE<a class=headerlink href=#2-setup-renee title="Permanent link">&para;</a></h2> <p><em>Estimated Reading Time: 3 Mintutes</em></p> <p>RENEE has two dependencies: <code>singularity</code> and <code>snakemake</code>. These dependencies can be installed by a sysadmin; however, snakemake is readily available through conda. Before running the pipeline or any of the commands below, please ensure singularity and snakemake are in your <code>$PATH</code>. Please see follow the instructions below for getting started with the RENEE pipeline.</p> <h3 id=21-login-to-cluster>2.1 Login to cluster<a class=headerlink href=#21-login-to-cluster title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 0.) ssh into cluster&#39;s head node</span>
<span class=c1># example below for Biowulf cluster</span>
ssh<span class=w> </span>-Y<span class=w> </span><span class=nv>$USER</span>@biowulf.nih.gov
</code></pre></div> <h3 id=22-grab-an-interactive-node>2.2 Grab an interactive node<a class=headerlink href=#22-grab-an-interactive-node title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 1.) Please do not run RENEE on the head node!</span>
<span class=c1># Grab an interactive node first</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>12</span>:00:00<span class=w> </span>-p<span class=w> </span>interactive<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>4</span><span class=w> </span>--pty<span class=w> </span>bash
</code></pre></div> <h3 id=23-load-dependencies>2.3 Load dependencies<a class=headerlink href=#23-load-dependencies title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Setup Step 2.) Add singularity and snakemake executables to $PATH</span>
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
</code></pre></div> <h2 id=3-building-reference-files>3. Building Reference files<a class=headerlink href=#3-building-reference-files title="Permanent link">&para;</a></h2> <p>In this example, we will start off by building reference files downloaded from <a href=https://www.gencodegenes.org/ >GENCODE</a>. We recommend downloading the <code>PRI</code> Genome FASTA file and annotation from <a href=https://www.gencodegenes.org/ >GENCODE</a>. These <code>PRI</code> reference files contain the primary chromosomes and scaffolds. We <strong>do not</strong> recommend downloading the <code>CHR</code> reference files!</p> <p>Checkout <a href=../Resources/ >this</a> list for currently available resources on Biowulf. If your required <strong>genome + annotation combination</strong> is NOT available, only then proceed to building your own reference files. Also, if you think that your <strong>genome + annotation combination</strong> may be beneficial for other Biowulf users of RENEE as well, then please request it to be added to RENEE's default resources by <a href=https://github.com/CCBR/RENEE/issues>opening an issue on Github</a>.</p> <h3 id=31-download-references-from-gencode>3.1 Download References from GENCODE<a class=headerlink href=#31-download-references-from-gencode title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Build Step 0.) Please do not run RENEE on the head node!</span>
<span class=c1># Grab an interactive node first</span>
<span class=c1># Assumes that you have already ssh-ed into cluster</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>12</span>:00:00<span class=w> </span>-p<span class=w> </span>interactive<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>4</span><span class=w> </span>--pty<span class=w> </span>bash

<span class=c1># Build Step 1.) Download the PRI Genome FASTA file for GRCh38.p13</span>
wget<span class=w> </span>ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/GRCh38.primary_assembly.genome.fa.gz
gzip<span class=w> </span>-d<span class=w> </span>GRCh38.primary_assembly.genome.fa.gz

<span class=c1># Build Step 2.) Download the PRI release 36 annotation</span>
wget<span class=w> </span>ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.primary_assembly.annotation.gtf.gz
gzip<span class=w> </span>-d<span class=w> </span>gencode.v36.primary_assembly.annotation.gtf.gz
</code></pre></div> <h3 id=32-run-build-pipeline>3.2 Run Build pipeline<a class=headerlink href=#32-run-build-pipeline title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Build Step 3.) Load dependencies</span>
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner

<span class=c1># Build Step 4.) Dry-run the build pipeline</span>
renee<span class=w> </span>build<span class=w>  </span>--ref-fa<span class=w> </span>GRCh38.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--ref-name<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--ref-gtf<span class=w> </span>gencode.v36.primary_assembly.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--gtf-ver<span class=w> </span><span class=m>36</span><span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/hg38_36<span class=w> </span>--dry-run


<span class=c1># Build Step 5.) Submit the build pipeline to cluster</span>
renee<span class=w> </span>build<span class=w>  </span>--ref-fa<span class=w> </span>GRCh38.primary_assembly.genome.fa<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--ref-name<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--ref-gtf<span class=w> </span>gencode.v36.primary_assembly.annotation.gtf<span class=w> </span><span class=se>\</span>
<span class=w>                </span>--gtf-ver<span class=w> </span><span class=m>36</span><span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/hg38_36
</code></pre></div> <p>An email notification will be sent out when the pipeline starts and ends. Once the build pipeline completes, you can run RENEE with the provided test dataset. Please see the instructions below for more information.</p> <h2 id=4-running-renee>4. Running RENEE<a class=headerlink href=#4-running-renee title="Permanent link">&para;</a></h2> <p>Run RENEE with the reference files we built above using hg38 (GRCh38.p13) Genome FASTA file and GENCODE release 36 annotation (GTF). For more information about how the reference files we generated, please see the instructions above. You can use those instructions as a guide for building any new reference genomes in the future.</p> <h3 id=41-dry-run-pipeline>4.1 Dry-run pipeline<a class=headerlink href=#41-dry-run-pipeline title="Permanent link">&para;</a></h3> <p>Dry-run the pipeline prior to submitting the pipeline's master job. Please note that if you wish to run RENEE with a new dataset, you will only need to update the values provided to the <code>--input</code> and <code>--output</code> arguments (and maybe <code>--genome</code>). The <code>--input</code> argument supports globbing. If this is the first time running RENEE with for given dataset, the <code>--output</code> directory should <em><strong>not</strong></em> exist on your local filesystem. It will be created automatically during runtime.</p> <div class=highlight><pre><span></span><code><span class=c1># Run Step 0.) Please do not run RENEE on the head node!</span>
<span class=c1># Grab an interactive node first</span>
<span class=c1># Assumes that you have already ssh-ed into cluster</span>
srun<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>12</span>:00:00<span class=w> </span>-p<span class=w> </span>interactive<span class=w> </span>--mem<span class=o>=</span>8gb<span class=w>  </span>--cpus-per-task<span class=o>=</span><span class=m>4</span><span class=w> </span>--pty<span class=w> </span>bash

<span class=c1># Run Step 1.) Load dependencies</span>
module<span class=w> </span>purge
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner

<span class=c1># Run Step 2.) Dry-run the pipeline with test dataset</span>
<span class=c1># And reference genome generated in the steps above</span>
<span class=c1># Test data consists of sub sampled FastQ files</span>
renee<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--input<span class=w> </span><span class=si>${</span><span class=nv>RENEE_HOME</span><span class=si>}</span>/.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--output<span class=w> </span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/runner_hg38_36/<span class=w>  </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/hg38_36/hg38_36.json<span class=w>  </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--star-2-pass-basic<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--dry-run
</code></pre></div> <h3 id=42-run-pipeline>4.2 Run pipeline<a class=headerlink href=#42-run-pipeline title="Permanent link">&para;</a></h3> <p>Kick off the pipeline by submitting the master job to the cluster. It is essentially the same command above without the <code>--dry-run</code> flag.</p> <div class=highlight><pre><span></span><code><span class=c1># Run Step 3.) Submit the master job</span>
<span class=c1># Runs the RENEE pipeline  with the</span>
<span class=c1># reference genome generated in the steps above</span>
<span class=c1># and with the test dataset</span>
renee<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--input<span class=w> </span><span class=si>${</span><span class=nv>RENEE_HOME</span><span class=si>}</span>/.tests/*.R?.fastq.gz<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--output<span class=w> </span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/runner_hg38_36/<span class=w>  </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/hg38_36/hg38_36.json<span class=w>  </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--star-2-pass-basic<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--dry-run
</code></pre></div> <p>An email notification will be sent out when the pipeline starts and ends.</p> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/RENEE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/RENEE</a> </span> <span><a href=../Theory/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../run/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 