<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="RENEE is an open-source, reproducible, and scalable best practices pipeline for analyzing RNA-sequencing data."><meta name=author content=CCBR><link href=../gui/ rel=prev><link href=../Resources/ rel=next><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.5.46"><title>Expected Output - RENEE Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.6f8fc17f.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#folder-details-and-file-descriptions class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="RENEE Documentation" class="md-header__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> RENEE Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Expected Output </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../Theory/ class=md-tabs__link> Theory </a> </li> <li class=md-tabs__item> <a href=../TLDR-RNA-seq/ class=md-tabs__link> Getting started </a> </li> <li class=md-tabs__item> <a href=../run/ class=md-tabs__link> Command Line </a> </li> <li class=md-tabs__item> <a href=../gui/ class=md-tabs__link> Graphical Interface </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=./ class=md-tabs__link> Expected Output </a> </li> <li class=md-tabs__item> <a href=../Resources/ class=md-tabs__link> Resources </a> </li> <li class=md-tabs__item> <a href=../../general-questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="RENEE Documentation" class="md-nav__button md-logo" aria-label="RENEE Documentation" data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> RENEE Documentation </label> <div class=md-nav__source> <a href=https://github.com/CCBR/RENEE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> CCBR/RENEE </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> About </span> </a> </li> <li class=md-nav__item> <a href=../Theory/ class=md-nav__link> <span class=md-ellipsis> Theory </span> </a> </li> <li class=md-nav__item> <a href=../TLDR-RNA-seq/ class=md-nav__link> <span class=md-ellipsis> Getting started </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> Command Line </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Command Line </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../run/ class=md-nav__link> <span class=md-ellipsis> run </span> </a> </li> <li class=md-nav__item> <a href=../build/ class=md-nav__link> <span class=md-ellipsis> build </span> </a> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> <span class=md-ellipsis> cache </span> </a> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> <span class=md-ellipsis> unlock </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../gui/ class=md-nav__link> <span class=md-ellipsis> Graphical Interface </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Expected Output </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Expected Output </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#folder-details-and-file-descriptions class=md-nav__link> <span class=md-ellipsis> Folder details and file descriptions </span> </a> <nav class=md-nav aria-label="Folder details and file descriptions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-bams class=md-nav__link> <span class=md-ellipsis> 1. bams </span> </a> </li> <li class=md-nav__item> <a href=#2-config class=md-nav__link> <span class=md-ellipsis> 2. config </span> </a> </li> <li class=md-nav__item> <a href=#3-deg_all class=md-nav__link> <span class=md-ellipsis> 3. DEG_ALL </span> </a> </li> <li class=md-nav__item> <a href=#4-fqscreen-and-fqscreen2 class=md-nav__link> <span class=md-ellipsis> 4. FQScreen and FQScreen2 </span> </a> </li> <li class=md-nav__item> <a href=#5-fusions class=md-nav__link> <span class=md-ellipsis> 5. fusions </span> </a> </li> <li class=md-nav__item> <a href=#6-kraken class=md-nav__link> <span class=md-ellipsis> 6. kraken </span> </a> </li> <li class=md-nav__item> <a href=#7-logfiles class=md-nav__link> <span class=md-ellipsis> 7. logfiles </span> </a> </li> <li class=md-nav__item> <a href=#8-nciccbr class=md-nav__link> <span class=md-ellipsis> 8. nciccbr </span> </a> </li> <li class=md-nav__item> <a href=#9-preseq class=md-nav__link> <span class=md-ellipsis> 9. preseq </span> </a> </li> <li class=md-nav__item> <a href=#10-qc-and-rawqc class=md-nav__link> <span class=md-ellipsis> 10. QC and rawQC </span> </a> </li> <li class=md-nav__item> <a href=#11-qualimap class=md-nav__link> <span class=md-ellipsis> 11. QualiMap </span> </a> </li> <li class=md-nav__item> <a href=#12-reports class=md-nav__link> <span class=md-ellipsis> 12. Reports </span> </a> </li> <li class=md-nav__item> <a href=#13-resources class=md-nav__link> <span class=md-ellipsis> 13. resources </span> </a> </li> <li class=md-nav__item> <a href=#14-rseqc class=md-nav__link> <span class=md-ellipsis> 14. RSeQC </span> </a> </li> <li class=md-nav__item> <a href=#15-star_files class=md-nav__link> <span class=md-ellipsis> 15. STAR_files </span> </a> </li> <li class=md-nav__item> <a href=#16-trim class=md-nav__link> <span class=md-ellipsis> 16. trim </span> </a> </li> <li class=md-nav__item> <a href=#17-workflow class=md-nav__link> <span class=md-ellipsis> 17. workflow </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../Resources/ class=md-nav__link> <span class=md-ellipsis> Resources </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex=0> <span class=md-ellipsis> FAQ </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../general-questions/ class=md-nav__link> <span class=md-ellipsis> General Questions </span> </a> </li> <li class=md-nav__item> <a href=../../troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> <li class=md-nav__item> <a href=../../contributing/ class=md-nav__link> <span class=md-ellipsis> How to contribute </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1>Expected Output</h1> <p>After a successful <code>renee</code> run execution for multisample paired-end data, the following files and folders are created in the output folder.</p> <div class=highlight><pre><span></span><code>renee_output/
├──<span class=w> </span>bams
├──<span class=w> </span>config<span class=w> </span>
├──<span class=w> </span>config.json<span class=w> </span><span class=c1># Contains the configuration and parameters used for this specific RENEE run</span>
├──<span class=w> </span>DEG_ALL
├──<span class=w> </span>dryrun.<span class=o>{</span>datetime<span class=o>}</span>.log<span class=w> </span><span class=c1># Output from the dry-run of the pipeline</span>
├──<span class=w> </span>FQscreen
├──<span class=w> </span>FQscreen2
├──<span class=w> </span>fusions
├──<span class=w> </span>kraken
├──<span class=w> </span>logfiles<span class=w> </span>
├──<span class=w> </span>nciccbr<span class=w> </span>
├──<span class=w> </span>preseq<span class=w> </span>
├──<span class=w> </span>QC
├──<span class=w> </span>QualiMap
├──<span class=w> </span>rawQC
├──<span class=w> </span>Reports
├──<span class=w> </span>resources
├──<span class=w> </span>RSeQC
├──<span class=w> </span>sample1.R1.fastq.gz<span class=w> </span>-&gt;<span class=w> </span>/path/to/input/fastq/files/sample1.R1.fastq.gz
├──<span class=w> </span>sample1.R2.fastq.gz<span class=w> </span>-&gt;<span class=w> </span>/path/to/input/fastq/files/sample1.R2.fastq.gz
...
..
.
├──<span class=w> </span>sampleN.R1.fastq.gz<span class=w> </span>-&gt;<span class=w> </span>/path/to/input/fastq/files/sampleN.R1.fastq.gz
├──<span class=w> </span>sampleN.R2.fastq.gz<span class=w> </span>-&gt;<span class=w> </span>/path/to/input/fastq/files/sampleN.R2.fastq.gz
├──<span class=w> </span>STAR_files
├──<span class=w> </span>trim
└──<span class=w> </span>workflow
</code></pre></div> <h2 id=folder-details-and-file-descriptions>Folder details and file descriptions<a class=headerlink href=#folder-details-and-file-descriptions title="Permanent link">&para;</a></h2> <h3 id=1-bams>1. <code>bams</code><a class=headerlink href=#1-bams title="Permanent link">&para;</a></h3> <p>Contains the STAR aligned reads for each sample analyzed in the run.</p> <div class=highlight><pre><span></span><code>/bams/
├──<span class=w> </span>sample1.fwd.bw<span class=w> </span><span class=c1># forward strand bigwig files suitable for a genomic track viewer like IGV</span>
├──<span class=w> </span>sample1.rev.bw<span class=w> </span><span class=c1># reverse strand bigwig files </span>
├──<span class=w> </span>sample1.p2.Aligned.toTranscriptome.out.bam<span class=w> </span><span class=c1># BAM alignments to transcriptome using STAR in two-pass mode</span>
├──<span class=w> </span>sample1.star_rg_added.sorted.dmark.bam<span class=w> </span><span class=c1># Read groups added and duplicates marked genomic BAM file (using STAR in two-pass mode)</span>
├──<span class=w> </span>sample1.star_rg_added.sorted.dmark.bam.bai
...
..
.
</code></pre></div> <h3 id=2-config>2. <code>config</code><a class=headerlink href=#2-config title="Permanent link">&para;</a></h3> <p>Contains config files for the pipeline.</p> <h3 id=3-deg_all>3. <code>DEG_ALL</code><a class=headerlink href=#3-deg_all title="Permanent link">&para;</a></h3> <p>Contains the output from RSEM estimating gene and isoform expression levels for each sample and also combined data matrix with all samples.</p> <div class=highlight><pre><span></span><code>/DEG_ALL/
├──<span class=w> </span>combined_TIN.tsv<span class=w> </span><span class=c1># RSeQC logfiles containing transcript integrity number information for all samples</span>
├──<span class=w> </span>RSEM.genes.expected_count.all_samples.txt<span class=w> </span><span class=c1># Expected gene counts matrix for all samples (useful for downstream differential expression analysis)</span>
├──<span class=w> </span>RSEM.genes.expected_counts.all_samples.reformatted.tsv<span class=w> </span><span class=c1># Expected gene counts matrix for all samples with reformatted gene symbols (format: ENSEMBLID | GeneName)</span>
├──<span class=w> </span>RSEM.genes.FPKM.all_samples.txt<span class=w> </span><span class=c1># FPKM Normalized expected gene counts matrix for all samples </span>
├──<span class=w> </span>RSEM.genes.TPM.all_samples.txt<span class=w> </span><span class=c1># TPM Normalized expected gene counts matrix for all samples</span>
├──<span class=w> </span>RSEM.isoforms.expected_count.all_samples.txt<span class=w> </span><span class=c1># File containing isoform level expression estimates for all samples.</span>
├──<span class=w> </span>RSEM.isoforms.FPKM.all_samples.txt<span class=w> </span><span class=c1># FPKM Normalized expected isoform counts matrix for all samples </span>
├──<span class=w> </span>RSEM.isoforms.TPM.all_samples.txt<span class=w> </span><span class=c1># TPM Normalized expected isoform counts matrix for all samples</span>
├──<span class=w> </span>sample1.RSEM.genes.results<span class=w> </span><span class=c1># Expected gene counts for sample 1</span>
├──<span class=w> </span>sample1.RSEM.isoforms.results<span class=w> </span><span class=c1># Expected isoform counts for sample 1</span>
├──<span class=w> </span>sample1.RSEM.stat<span class=w> </span><span class=c1># RSEM stats for sample 1</span>
│<span class=w>   </span>├──<span class=w> </span>sample1.RSEM.cnt<span class=w> </span>
│<span class=w>   </span>├──<span class=w> </span>sample1.RSEM.model
│<span class=w>   </span>└──<span class=w> </span>sample1.RSEM.theta
├──<span class=w> </span>sample1.RSEM.time<span class=w> </span><span class=c1># Run time log for sample 1</span>
...
..
.
├──<span class=w> </span>sampleN.RSEM.genes.results
├──<span class=w> </span>sampleN.RSEM.isoforms.results
├──<span class=w> </span>sampleN.RSEM.stat
│<span class=w>   </span>├──<span class=w> </span>sampleN.RSEM.cnt
│<span class=w>   </span>├──<span class=w> </span>sampleN.RSEM.model
│<span class=w>   </span>└──<span class=w> </span>sampleN.RSEM.theta
└──<span class=w> </span>sampleN.RSEM.time
</code></pre></div> <h3 id=4-fqscreen-and-fqscreen2>4. <code>FQScreen</code> and <code>FQScreen2</code><a class=headerlink href=#4-fqscreen-and-fqscreen2 title="Permanent link">&para;</a></h3> <p>These folders contain results from quality-control step to screen for different sources of contamination. FastQ Screen compares your sequencing data to a set of different reference genomes to determine if there is contamination. It allows a user to see if the composition of your library matches what you expect. These results are plotted in the multiQC report.</p> <h3 id=5-fusions>5. <code>fusions</code><a class=headerlink href=#5-fusions title="Permanent link">&para;</a></h3> <p>Contains gene fusions output for each sample.</p> <div class=highlight><pre><span></span><code>fusions/
├──<span class=w> </span>sample1_fusions.arriba.pdf
├──<span class=w> </span>sample1_fusions.discarded.tsv<span class=w> </span><span class=c1># Contains all events that Arriba classified as an artifact or that are also observed in healthy tissue. </span>
├──<span class=w> </span>sample1_fusions.tsv<span class=w> </span><span class=c1># Contains fusions for sample 1 which pass all of Arriba&#39;s filters. The predictions are listed from highest to lowest confidence. </span>
├──<span class=w> </span>sample1.p2.arriba.Aligned.sortedByCoord.out.bam<span class=w> </span><span class=c1># Sorted BAM file for Arriba&#39;s Visualization</span>
├──<span class=w> </span>sample1.p2.arriba.Aligned.sortedByCoord.out.bam.bai
├──<span class=w> </span>sample1.p2.Log.final.out<span class=w> </span><span class=c1># STAR final log file</span>
├──<span class=w> </span>sample1.p2.Log.out<span class=w> </span><span class=c1># STAR runtime log file</span>
├──<span class=w> </span>sample1.p2.Log.progress.out<span class=w> </span><span class=c1># log files</span>
├──<span class=w> </span>sample1.p2.Log.std.out<span class=w> </span><span class=c1># STAR runtime output log</span>
├──<span class=w> </span>sample1.p2.SJ.out.tab<span class=w> </span><span class=c1>#  Summarizes the high confidence splice junctions for sample 1</span>
├──<span class=w> </span>sample1.p2._STARgenome<span class=w> </span><span class=c1># Extra files generated during STAR aligner </span>
│<span class=w>   </span>├──<span class=w> </span>exonGeTrInfo.tab
│<span class=w>   </span>├──<span class=w> </span>.
│<span class=w>   </span>├──<span class=w> </span>.
│<span class=w>   </span>└──<span class=w> </span>transcriptInfo.tab<span class=w> </span>
├──<span class=w> </span>sample1.p2._STARpass1<span class=w> </span><span class=c1># Extra files generated during STAR first pass </span>
│<span class=w>   </span>├──<span class=w> </span>.
│<span class=w>   </span>└──<span class=w> </span>.
...
..
.
</code></pre></div> <h3 id=6-kraken>6. <code>kraken</code><a class=headerlink href=#6-kraken title="Permanent link">&para;</a></h3> <p>Contains per sample kraken output files which is a Quality-control step to assess for potential sources of microbial contamination. Kraken is used in conjunction with Krona to produce an interactive reports stored in <code>.krona.html</code> files. These results are present in the multiQC report.</p> <h3 id=7-logfiles>7. <code>logfiles</code><a class=headerlink href=#7-logfiles title="Permanent link">&para;</a></h3> <p>Contains logfiles for the entire RENEE run, job error/output files for each individual job that was submitted to SLURM, and some other stats generated by different software. Important to diagnose errors if the pipeline fails. The per sample stats information is present in the mulitQC report. </p> <div class=highlight><pre><span></span><code>/logfiles/
├──<span class=w> </span>master.log<span class=w> </span><span class=c1># Logfile for the main (master) RENEE job</span>
├──<span class=w> </span>mjobid.log<span class=w> </span><span class=c1># SLURM JOBID for the master RENEE job</span>
├──<span class=w> </span>runtime_statistics.json<span class=w> </span><span class=c1># Runtime statistics for each rule in the RENEE run</span>
├──<span class=w> </span>sample1.flagstat.concord.txt<span class=w> </span><span class=c1># sample mapping stats</span>
├──<span class=w> </span>sample1.p2.Log.final.out<span class=w> </span><span class=c1># sample STAR alignment stats</span>
├──<span class=w> </span>sample1.RnaSeqMetrics.txt<span class=w> </span><span class=c1># sample stats collected by Picard CollectRnaSeqMetrics</span>
├──<span class=w> </span>sample1.star.duplic<span class=w> </span><span class=c1># Mark duplicate metrics</span>
...
..
.
├──<span class=w> </span>slurmfiles
│<span class=w>   </span>├──<span class=w> </span><span class=o>{</span>MASTER_JOBID<span class=o>}</span>.<span class=o>{</span>JOBID<span class=o>}</span>.<span class=o>{</span>rule<span class=o>}</span>.<span class=o>{</span>wildcards<span class=o>}</span>.out
│<span class=w>   </span>├──<span class=w> </span><span class=o>{</span>MASTER_JOBID<span class=o>}</span>.<span class=o>{</span>JOBID<span class=o>}</span>.<span class=o>{</span>rule<span class=o>}</span>.<span class=o>{</span>wildcards<span class=o>}</span>.err
│<span class=w>   </span>...
│<span class=w>   </span>..
│<span class=w>   </span>.
├──<span class=w> </span>snakemake.log<span class=w> </span><span class=c1># The snakemake log file which documents the entire pipeline log</span>
├──<span class=w> </span>snakemake.log.jobby<span class=w> </span><span class=c1># Detailed summary report for each individual job. </span>
└──<span class=w> </span>snakemake.log.jobby.short<span class=w> </span><span class=c1># Short summary report for each individual job. </span>
</code></pre></div> <h3 id=8-nciccbr>8. <code>nciccbr</code><a class=headerlink href=#8-nciccbr title="Permanent link">&para;</a></h3> <p>Contain Arriba resources for gene fusion estimation. Manually curated and only exist for a few reference genomes (mm10, hg38, hg19).</p> <h3 id=9-preseq>9. <code>preseq</code><a class=headerlink href=#9-preseq title="Permanent link">&para;</a></h3> <p>Contains library complexity curves for each sample. These results are part of the multiQC report.</p> <h3 id=10-qc-and-rawqc>10. <code>QC</code> and <code>rawQC</code><a class=headerlink href=#10-qc-and-rawqc title="Permanent link">&para;</a></h3> <p>Contains per sample output from FastQC for raw and adapter trimmed fastq files with insert size estimates. These results are part of the multiQC report.</p> <h3 id=11-qualimap>11. <code>QualiMap</code><a class=headerlink href=#11-qualimap title="Permanent link">&para;</a></h3> <p>Contains per sample output for Quality-control step to assess various post-alignment metrics and a secondary method to calculate insert size. These results are part of the multiQC report.</p> <h3 id=12-reports>12. <code>Reports</code><a class=headerlink href=#12-reports title="Permanent link">&para;</a></h3> <p>Contains the multiQC report which visually summarizes the quality control metrics and other statistics for each sample (<code>multiqc_report.html</code>). All the data tables used to generate the multiQC report is available in the <code>multiqc_data</code> folder. The <code>RNA_report.html</code> file is an interactive report the aggregates sample quality-control metrics across all samples. This interactive report to allow users to identify problematic samples prior to downstream analysis. It uses flowcell and lane information from the FastQ file.</p> <h3 id=13-resources>13. <code>resources</code><a class=headerlink href=#13-resources title="Permanent link">&para;</a></h3> <p>Contains resources necessary to run the RENEE pipeline.</p> <h3 id=14-rseqc>14. <code>RSeQC</code><a class=headerlink href=#14-rseqc title="Permanent link">&para;</a></h3> <p>Contains various QC metrics for each sample collected by RSeQC. These results are part of the multiQC report.</p> <h3 id=15-star_files>15. <code>STAR_files</code><a class=headerlink href=#15-star_files title="Permanent link">&para;</a></h3> <p>Contains log files, splice junction tab file (<code>SJ.out.tab</code>), and <code>ReadsPerGene.out.tab</code> file, and other various output files for each sample generated by STAR aligner.</p> <h3 id=16-trim>16. <code>trim</code><a class=headerlink href=#16-trim title="Permanent link">&para;</a></h3> <p>Contains adapter trimmed FASTQ files for each sample used for all the downstream analysis.</p> <div class=highlight><pre><span></span><code>trim
├──<span class=w> </span>sample1.R1.trim.fastq.gz
├──<span class=w> </span>sample1.R2.trim.fastq.gz
...
..
.
├──<span class=w> </span>sampleN.R1.trim.fastq.gz
└──<span class=w> </span>sampleN.R2.trim.fastq.gz
</code></pre></div> <h3 id=17-workflow>17. <code>workflow</code><a class=headerlink href=#17-workflow title="Permanent link">&para;</a></h3> <p>Contains the RENEE pipeline workflow.</p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> 2024-08-22 </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2023 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160m368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160M0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7M405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.4zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0m-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.3c73.7 0 133.4 59.7 133.4 133.3 0 14.7-11.9 26.7-26.7 26.7H154.6c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.83f73b43.min.js></script> </body> </html>